crates/nu-command/src/filters/default.rs

```diff
@@ -110,70 +139,150 @@ fn default(
     stack: &mut Stack,
     call: &Call,
     input: PipelineData,
+    default_value: Value,
     default_when_empty: bool,
+    columns: &[String],
 ) -> Result<PipelineData, ShellError> {
+    let input_span = input.span().unwrap_or_else(Span::unknown);
+    let mut default_value = DefaultValue::new(engine_state, stack, default_value);
     let metadata = input.metadata();
-    let value: Value = call.req(engine_state, stack, 0)?;
-    let column: Option<Spanned<String>> = call.opt(engine_state, stack, 1)?;
-
-    if let Some(column) = column {
-        input
-            .map(
-                move |mut item| match item {
-                    Value::Record {
-                        val: ref mut record,
-                        ..
-                    } => {
-                        let record = record.to_mut();
-                        if let Some(val) = record.get_mut(&column.item) {
-                            if matches!(val, Value::Nothing { .. })
-                                || (default_when_empty && val.is_empty())
-                            {
-                                *val = value.clone();
-                            }
-                        } else {
-                            record.push(column.item.clone(), value.clone());
-                        }
-
-                        item
-                    }
-                    _ => item,
-                },
-                engine_state.signals(),
-            )
-            .map(|x| x.set_metadata(metadata))
+
+    // If user supplies columns, check if input is a record or list of records
+    // and set the default value for the specified record columns
+    if !columns.is_empty() {
+        if matches!(input, PipelineData::Value(Value::Record { .. }, _)) {
+            let Value::Record {
+                val: ref mut record,
+                internal_span,
+            } = input.into_value(input_span)?
+            else {
+                unreachable!()
+            };
+            let record = record.to_mut().into_spanned(internal_span);
+            fill_record(record, &mut default_value, columns, default_when_empty)
+                .map(|x| x.into_pipeline_data().set_metadata(metadata))
```

Comment:
```suggestion
            let record = input.into_value(input_span)?.into_record()?;
            fill_record(
                record,
                input_span,
                &mut default_value,
                columns.as_slice(),
                default_when_empty,
            )
            .map(|x| x.into_pipeline_data_with_metadata(metadata))
```

Since we're already coercing this into a concrete `Value`, we can just extract the `Record` value directly rather than mutating it by reference, which is a bit nicer to work with